vignettes/minimal-rstan-package.Rmd

```diff
@@ -0,0 +1,160 @@
+---
+title: "A working example for writing a package that depends on RStan"
+author: "Stefan Siegert"
+date: "`r Sys.Date()`"
+output:
+  rmarkdown::html_vignette:
+    toc: true
+vignette: >
+  %\VignetteIndexEntry{Minimal RStan Package}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, echo=FALSE, eval=TRUE}
+knitr::opts_chunk$set(eval=FALSE)
+```
+
+## Creating a minimal package that uses Stan, step-by-step
+
+This vignette is a step-by-step guide to create a functioning R package that uses Stan.
+It is complementary to the [rstantools developer guidelines vignette](developer-guidelines.html) which you should read first.
+
+
+To start off, we use `rstan_package_skeleton` to initialise a bare-bones package directory.
+The name of our package will be `rstanlm`; it will fit a simple linear regression model using Stan.
+Under the hood, the function `rstan_package_skeleton` calls `utils::package.skeleton()` which won't allow creating a completely empty package directory, so we initialise it with a dummy variable which we will delete later on:
+
+```{r}
+library(rstantools)
+rstan_package_skeleton(name='rstanlm', environment=as.environment(list(dummy=42)))
+```
+
+The newly created package directory is called `rstanlm` and we change to it to make a few modifications:
+
+```{r}
+setwd('rstanlm')
+```
+
+We remove the following files, which are either unneeded, or will be automatically recreated later on:
+
+```{r}
+file.remove('Read-and-delete-me')
+file.remove('data/dummy.rda')
+file.remove('man/dummy.Rd')
+file.remove('NAMESPACE')
+file.remove('man/rstanlm-package.Rd')
+```
+
+Some of the magic to make Stan models available as part of your package happens in the file `cleanup` (or `cleanup.win` on Windows).
+The `cleanup` script is executed during the build process, but for this to work its file mode has to be changed to executable:
+
+```{r}
+Sys.chmod('cleanup', mode='755')
+Sys.chmod('cleanup.win', mode='755')
+```
+
+Our package is going to use Stan to fit a simple linear regression model to a variable `y` using `x` as a predictor.
+We write the following Stan model to the file `exec/lm.stan`:
+
+```
+data {
+  int<lower=1> N;
+  vector[N] x;
+  vector[N] y;
+}
+parameters {
+  real intercept;
+  real beta;
+  real<lower=0> sigma;
+}
+model {
+  y ~ normal(intercept + beta * x, sigma);
+}
+```
+
+The `exec` subdirectory can contain further Stan models. 
+During installation, all Stan models will be compiled and saved in the list `stanmodels` that can then be used by R function in the package.
+The rule is that the Stan model compiled from the model code in `exec/foo.stan` is stored as list element `stanmodels$foo`.
+
+We next create the file `R/lm_stan.R` where we define the function `lm_stan` in which our compiled Stan model is being used.
+A comment block in roxygen2 syntax ensures that the function has a help file and that it is added to the NAMESPACE:
```

Comment:
Can you make 'roxygen2' a hyperlink: `[roxygen2](https://CRAN.R-project.org/package=roxygen2)`